src/EventListener/PackageListener.php

```diff
@@ -0,0 +1,96 @@
+<?php declare(strict_types=1);
+
+/*
+ * This file is part of Packagist.
+ *
+ * (c) Jordi Boggiano <j.boggiano@seld.be>
+ *     Nils Adermann <naderman@naderman.de>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace App\EventListener;
+
+use App\Entity\AuditRecord;
+use App\Entity\Package;
+use App\Entity\User;
+use App\Util\DoctrineTrait;
+use Doctrine\Bundle\DoctrineBundle\Attribute\AsEntityListener;
+use Doctrine\DBAL\Connection;
+use Doctrine\ORM\EntityManager;
+use Doctrine\Persistence\Event\LifecycleEventArgs;
+use Doctrine\ORM\Event\PreUpdateEventArgs;
+use Doctrine\Persistence\ManagerRegistry;
+use Predis\Client;
+use Symfony\Component\Security\Core\Security;
+
+#[AsEntityListener(event: 'postPersist', entity: Package::class)]
+#[AsEntityListener(event: 'preRemove', entity: Package::class)]
+#[AsEntityListener(event: 'preUpdate', entity: Package::class)]
+#[AsEntityListener(event: 'postUpdate', entity: Package::class)]
+class PackageListener
+{
+    use DoctrineTrait;
+
+    /** @var list<AuditRecord> */
+    private array $buffered = [];
+
+    public function __construct(
+        private ManagerRegistry $doctrine,
+        private Security $security,
+    ) {
+    }
+
+    /**
+     * @param LifecycleEventArgs<EntityManager> $event
+     */
+    public function postPersist(Package $package, LifecycleEventArgs $event): void
+    {
+        $record = AuditRecord::packageCreated($package, $this->getUser());
+        $this->getEM()->persist($record);
+        $this->getEM()->flush();
+    }
+
+    /**
+     * @param LifecycleEventArgs<EntityManager> $event
+     */
+    public function preRemove(Package $package, LifecycleEventArgs $event): void
+    {
+        $record = AuditRecord::packageDeleted($package, $this->getUser());
+        $this->getEM()->persist($record);
+        // let the record be flushed together with the entity
+    }
+
+    /**
+     * @param PreUpdateEventArgs $event
+     */
+    public function preUpdate(Package $package, PreUpdateEventArgs $event): void
+    {
+        if ($event->hasChangedField('repository')) {
+            // buffering things to be flushed in postUpdate as flushing here results in an infinite loop
```

Comment:
couldn't you persist them here and let them be flushed together with the entity ?

Reply:
I thought so too but it didn't do anything..